File: src/entities/Deposit.ts
import { Ore, RawOre } from './Ore'
import { TResource } from './Resource'
import { ResourceAmount } from './static/ResourceAmount'
import { StaticDeposit, TStaticDeposit } from './static'

export type RawDeposit = {
  pid    :string
  sdid   :string

  ores :  RawOre[]
  rate        :number
  syncedAt   :number
}

export type TDeposit = {
  pid     :string
  oreStorages :Ore[]
  rate        :number
  syncedAt   :number
  base       : TStaticDeposit
  totalOres   : number
}

export class Deposit {
  planetId     :string
  oreStorages :Ore[]
  rate        :number
  syncedAt   :number
  base       : StaticDeposit
  totalOres   : number

  constructor(data: RawDeposit, sDeposit: StaticDeposit) {
    this.base = sDeposit
    this.planetId = data.pid
    this.totalOres = 0

    this.oreStorages = sDeposit.ores.map(bOre => {
      const rOre = data.ores.find(ore => ore.srid == bOre.resource.id)
      return new Ore(rOre || { srid: bOre.resource.id, amount: '0' }, bOre.resource)
    })
    this.rate = data.rate
    this.syncedAt = data.syncedAt
    this.totalOres = this.oreStorages.reduce((acc, ore) => acc + ore.amount, 0)
  }

  static initFromStatic(planetId: string, base: StaticDeposit, syncedAt: number): Deposit {
    return new Deposit({
      pid: planetId,
      rate: base.rate,
      sdid: base.id,
      ores: base.ores.map(ore => ({
        srid: ore.resource.id,
        amount: '0'
      })),
      syncedAt,
    }, base)
  }

  toRaw(): RawDeposit {
    return {
      ores: this.oreStorages.map(ore => ({
        srid: ore.id,
        amount: ore.amount.toString()
      })),
      rate: this.rate,

      pid    : this.planetId,
      sdid   : this.base.id,

      syncedAt   : this.syncedAt
    }
  }


  sync(ts: number): Deposit {
    if (ts <= this.syncedAt) return this
    const timeDiff = (ts - this.syncedAt)/1000
    let totalOres = 0
    this.oreStorages.forEach(ore => {
      const bOre = this.base.ores.find(bo => bo.resource.id == ore.base.id)
      if (!bOre) {
        console.error('Cannot find base Ore. Skip sync', ore)
        return this
      }
      const changes = timeDiff * bOre.ratio * this.rate / 100
      ore.amount += changes
      totalOres += changes
    })
    this.totalOres += totalOres
    this.syncedAt = ts
    return this
  }

  withdrawOres(capacity: number): ResourceAmount[] {
    const result: ResourceAmount[] = []
    for (let i = 0; i < this.base.ores.length && capacity > 0; i++) {
      const oreAtTs = this.oreStorages[i]
      const amount = Math.floor(oreAtTs.amount)
      const resource = new ResourceAmount({ srid: oreAtTs.base.id, amount: "0" }, oreAtTs.base)
      if (capacity > oreAtTs.amount) {
        resource.amount = BigInt(amount)
        capacity -= amount
        oreAtTs.amount -= amount
      } else if (capacity > 0) {
        resource.amount = BigInt(capacity)
        oreAtTs.amount -= capacity
        capacity = 0
      }
      this.totalOres = this.totalOres - Number(resource.amount)
      result.push(resource)
    }
    return result
  }
}

type DepositResource = Pick<TResource, 'base'> & { amount: number }

File: src/components/miner/deposit.rs
use std::ptr::NonNull;

use crate::{components::common::{Ore, ProductAmount}, entities::{base::RawBaseDeposit, raw::RawDeposit}};

struct Deposit {
  planet_id     :String,
  ore_storages :Vec<Ore>,
  rate        :f64,
  synced_at   :f64,
  base       : RawBaseDeposit,
  total_ores   : f64,
}

impl Deposit {
  fn sync(&mut self, ts: f64) {
    if ts <= self.synced_at {
        return;
    }
    let time_diff = (ts - self.synced_at) / 1000.0;
    let mut total_ores = 0.0;
    for ore in &mut self.ore_storages {
      let mut bore: Option<&crate::entities::base::Ore> = None;
      for b_ore in &self.base.ores {
        if ore.id == b_ore.srid {
          bore = Some(b_ore);
          break;
        }
      }
      if let Some(bore) = bore {
        let changes = time_diff * bore.ratio * self.rate / 100.0;
        ore.amount += changes;
        total_ores += changes;
      }

    }
    self.total_ores += total_ores;
    self.synced_at = ts;
  }

  fn withdraw_ores(deposit: &mut Deposit, capacity: f64) -> Vec<ProductAmount> {
    let mut result: Vec<ProductAmount> = Vec::new();
    let mut capacity = capacity;
    for i in 0..deposit.base.ores.len() {
        if capacity <= 0.0 {
            break;
        }
        let ore_at_ts = &mut deposit.ore_storages[i];
        let amount = ore_at_ts.amount.floor() as i64;
        // let product = ProductAmount {
        //     srid: ore_at_ts.id.clone(),
        //     amount: 0.into(),
        //     // base: ore_at_ts.base.clone(),
        // };
        if capacity > ore_at_ts.amount {
            // product.amount = amount.into();
            // capacity -= amount as f64;
            // ore_at_ts.amount -= amount as f64;
        } else if capacity > 0.0 {
            // product.amount = capacity as i64.into();
            ore_at_ts.amount -= capacity;
            capacity = 0.0;
        }
        // deposit.total_ores -= product.amount as f64;
        // result.push(product);
    }
    result
  }
}